### PR TITLE
Amqnet 723

### PR DESCRIPTION
**Pull request by lukeabsent**

### CODE REVIEW

**Comment by michaelandrepearce** [comment]
Can these be separate prs please

File: src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs [hunk 1]
```diff
@@ -162,55 +160,68 @@ private void SendSync(global::Amqp.Message message, DeliveryState deliveryState)
...
+            try
             {
-                cts = new CancellationTokenSource(TimeSpan.FromMilliseconds(session.Connection.Provider.SendTimeout));
-                cts.Token.Register(_ =>
-                {
-                    var timeoutException = ExceptionSupport.GetTimeoutException(this.senderLink, $"The operation did not complete within the allocated time {session.Connection.Provider.SendTimeout}ms.");
-                    tcs.TrySetException(timeoutException);
-                }, null);
+                outcome = await senderLink.SendAsync(message, deliveryState, session.Connection.Provider.SendTimeout).Await(); 
+                // NmsConnectionInfo.INFINITE is -1 which is the same as System.Threading.Timeout.Infinite that should have meaning of not start the internal timer in AmqpSendTask
```
Comment (michaelandrepearce) [inline src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:172]: Can we use the connection info still

**Comment by lukeabsent** [comment]
> 
> 
> Can these be separate prs please

723 was branched from 637, once 637 gets merged, I suppose there will be just 723 visible on the commit lists

File: test/Apache-NMS-AMQP-Test/Integration/Async/TransactionsIntegrationTestAsync.cs [hunk 2]
```diff
@@ -997,6 +998,12 @@ public async Task TestSendAfterCoordinatorLinkClosedDuringTX()
 
                 testPeer.WaitForAllMatchersToComplete(2000);
 
+                // Make sure that above line 'WaitForAllMatchersToComplete' also finishes on session/producer side
+                // In order for producer.Send to not really send, transaction on this end needs to be IsTransactionFailed
+                // in AmqpProducer  if (session.IsTransacted && session.IsTransactionFailed), and sometimes producer.Send tries
+                // to send before this flag is set, which would cause actual transfer
+                await Task.Delay(10);
```
Comment (michaelandrepearce) [inline test/Apache-NMS-AMQP-Test/Integration/Async/TransactionsIntegrationTestAsync.cs:1005]: This should test and wait of a condition or fail. Just random pause will mean test becomes temperamental on faster or slower machines

**Comment by michaelandrepearce** [comment]
There is no jiras too in here. Can we seperate those too please.

File: docs/configuration.md [hunk 3]
```diff
@@ -23,6 +23,7 @@ The options apply to the behavior of the NMS objects such as Connection, Session
 - **nms.requestTimeout** Timeout value that controls how long the client waits on completion of various synchronous interactions, such as opening a producer or consumer, before returning an error. Does not affect synchronous message sends. By default the client will wait indefinitely for a request to complete.
 - **nms.clientIdPrefix** Optional prefix value that is used for generated Client ID values when a new Connection is created for the JMS ConnectionFactory. The default prefix is 'ID:'.
 - **nms.connectionIdPrefix** Optional prefix value that is used for generated Connection ID values when a new Connection is created for the JMS ConnectionFactory. This connection ID is used when logging some information from the JMS Connection object so a configurable prefix can make breadcrumbing the logs easier. The default prefix is 'ID:'.
+- **nms.maxMessageSize** Optional, specifies maximum message size (body) to be accepted for send. If its bigger, exeception would be thrown. The default value is 0 which means no limit.
```
Comment (michaelandrepearce) [inline docs/configuration.md:26]: Can we please use -1 to mean no limt.

**Comment by michaelandrepearce** [comment]
@lukeabsent whats left on this one, to take it out of draft?

**Comment by lukeabsent** [comment]
@michaelandrepearce  I put PR to amqpnetlite that exposes method for estimating message size, it got merged. but its not released yet. well, we can discuss continuing without it and having own message body estimates.

**Comment by Havret** [comment]
That's pretty odd feature to add in a library code. I don't see anything along these lines in qpid-jms. Why do we need this?

File: src/NMS.AMQP/Meta/NmsConnectionInfo.cs [hunk 4]
```diff
@@ -62,6 +63,9 @@ public NmsConnectionInfo(NmsConnectionId connectionId)
         public string TopicPrefix { get; set; }
         public ushort ChannelMax { get; set; } = DEFAULT_CHANNEL_MAX;
         public int MaxFrameSize { get; set; } = DEFAULT_MAX_FRAME_SIZE;
+
+        public int? MaxMessageSize { get; set; } = DEFAULT_MAX_MESSAGE_SIZE;
```
Comment (michaelandrepearce) [inline src/NMS.AMQP/Meta/NmsConnectionInfo.cs:67]: why is this 'int?' why not just int, it has a default.

File: test/Apache-NMS-AMQP-Test/Integration/NMSConsumerIntegrationTest.cs [hunk 5]
```diff
@@ -1030,13 +1029,12 @@ public void TestReceiveBodyNoWait<T>(T inputValue)
                 // Release and allow link to flow
                 beforeFlow.Set();
                 // Give short time to arrive
-                Thread.Sleep(100);
+                Thread.Sleep(500);
```
Comment (michaelandrepearce) [inline test/Apache-NMS-AMQP-Test/Integration/NMSConsumerIntegrationTest.cs:1032]: why this change is needed? has this change caused significant perf degration?

File: src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs [hunk 6]
```diff
@@ -194,6 +196,31 @@ private async Task SendAsync(global::Amqp.Message message, DeliveryState deliver
...
+        {
+            switch (message.Body)
+            {
+                case String str:
+                    return str.Length;
+                case byte[] bytes:
+                    return bytes.Length;
+            }
+
+            return 64;
```
Comment (michaelandrepearce) [inline src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:222]: What? this seems very arbitary

Comment (michaelandrepearce) [reply src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:222]: surely a null body is 0 size. and all other bodys i would expect we have to handle. cannot just have this feature but only partial support.

**Comment by michaelandrepearce** [comment]
@lukeabsent see comments, also conflicts.

File: test/Apache-NMS-AMQP-Test/Integration/NMSConsumerIntegrationTest.cs (hunk 5 above)
Comment (lukeabsent) [reply test/Apache-NMS-AMQP-Test/Integration/NMSConsumerIntegrationTest.cs:1032]: there was one travis build test failure, probably message didnt manage to get to receiver and receivebodynowait returned default. unfortunately i dont know of the way to find out if it reached receiver before calling that method.

File: src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs (hunk 6 above)
Comment (lukeabsent) [reply src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:222]: this method is still very imperfect, its rather draft. 
this pr waits for amqpnetlite https://github.com/Azure/amqpnetlite/pull/472/files , but this one may also not be perfect message size, perhaps there would have to be some combined approach.

Comment (lukeabsent) [reply src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:222]: I would convert this pr into Draft, but I dont think I can do it. Probably you could do it @michaelandrepearce ?
https://github.blog/changelog/2020-04-08-convert-pull-request-to-draft/

Comment (Havret) [reply src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:222]: @lukeabsent converted to draft

Comment (lukeabsent) [reply src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:222]: @Havret thanks

File: docs/configuration.md [hunk 7]
```diff
@@ -24,6 +24,7 @@ The options apply to the behavior of the NMS objects such as Connection, Session
 - **nms.clientIdPrefix** Optional prefix value that is used for generated Client ID values when a new Connection is created for the JMS ConnectionFactory. The default prefix is 'ID:'.
 - **nms.connectionIdPrefix** Optional prefix value that is used for generated Connection ID values when a new Connection is created for the JMS ConnectionFactory. This connection ID is used when logging some information from the JMS Connection object so a configurable prefix can make breadcrumbing the logs easier. The default prefix is 'ID:'.
 - **nms.maxNewConnectionRatePerSec** Allowed approximated rate for how fast connection factory is allowed to create new connection. If there is more request, they will have to wait. Default value is -1 which means unlimited.
+- **nms.maxMessageSize** Specifies maximum message size (body) to be accepted for send. If its bigger, exception may be thrown. The default value is -1 which means no limit.
```
Comment (gemmellr) [inline docs/configuration.md:27]: I would call it outboundMaxMessageSize or similar if doing this (I dont see a need for it) since the change is specifically about sent message size.

File: src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs (hunk 6 above)
Comment (gemmellr) [reply src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:222]: The message estimate method from the linked PR is also a pretty terrible estimate:

```
        public int GetEstimatedMessageSize()
        {
            int size = 0;
            if (this.Header != null) size += 64;
            if (this.DeliveryAnnotations != null) size += 64;
            if (this.MessageAnnotations != null) size += 64;
            if (this.Properties != null) size += 64;
            if (this.ApplicationProperties != null) size += 64;
            if (this.BodySection != null) size += GetEstimatedBodySize(this.BodySection) + 8;
            if (this.Footer != null) size += 64;
            return size;
        }
```
I dont think Header can ever be close to 64 bytes, Properties could be bigger depending on particular fields values, the DeliveryAnnotations/MessageAnnotations/ApplicationProperties/Footer are essentially arbitrary maps that can be any size depending on their contents. Then of course theres the apparent fallback to 64 if it doesnt handle sizing the body type.

Comment (gemmellr) [reply src/NMS.AMQP/Provider/Amqp/AmqpProducer.cs:222]: The message estimate method from the linked PR is also a pretty terrible estimate:

```
        public int GetEstimatedMessageSize()
        {
            int size = 0;
            if (this.Header != null) size += 64;
            if (this.DeliveryAnnotations != null) size += 64;
            if (this.MessageAnnotations != null) size += 64;
            if (this.Properties != null) size += 64;
            if (this.ApplicationProperties != null) size += 64;
            if (this.BodySection != null) size += GetEstimatedBodySize(this.BodySection) + 8;
            if (this.Footer != null) size += 64;
            return size;
        }
```
I dont think Header can ever be close to 64 bytes, Properties could be bigger depending on particular fields values, the DeliveryAnnotations/MessageAnnotations/ApplicationProperties/Footer are essentially arbitrary maps that can be any size depending on their contents. Then of course theres the apparent fallback to 64 if it doesnt handle sizing the body type.

File: docs/configuration.md (hunk 7 above)
Comment (lukeabsent) [reply docs/configuration.md:27]: sure, we can try adjusting the name,
@gemmellr I was thinking instead of using GetEstimatedMessageSize there can be custom code for estimating body size

**Comment by michaelandrepearce** [comment]
@lukeabsent whats occuring with this, its been 2 yrs and you didnt resolve @gemmellr comments, are we closing this?

**Comment by lukeabsent** [comment]
yes @michaelandrepearce from my side I think we can ignore/close this pr
